Troy Morgan    
Week 8 Independent Lab    
BGEN 632, Graduate Introduction to Python    
Submitted April 15, 2025    

Independent Lab: Advanced Data Manipulation

In [1]:
## Imports
import os
import numpy as np
import pandas as pd

In [2]:
# Set working directory and confirm.
os.chdir("\\Users\\troy.morgan\\Documents\\GitHub\\week8labs/data")  
os.getcwd() 

'C:\\Users\\troy.morgan\\Documents\\GitHub\\week8labs\\data'

In [3]:
# Import data
gh_turnover_four_projects = pd.read_csv("gh_turnover_four_projects.csv")

# Quick inspection:
gh_turnover_four_projects

project_id     domain language  windows  window_idx  num_team  \
0          1101      OTHER   Python       23          22        26   
1          1101      OTHER   Python       23          22        26   
2          1101      OTHER   Python       23          22        26   
3          1101      OTHER   Python       23          22        26   
4          1101      OTHER   Python       23          22        26   
..          ...        ...      ...      ...         ...       ...   
468       24293  DEV_FRAME   Python       24          23        12   
469       24293  DEV_FRAME   Python       24          24        12   
470       24293  DEV_FRAME   Python       24          24        12   
471       24293  DEV_FRAME   Python       24          24        12   
472       24293  DEV_FRAME   Python       24          24        12   

     num_commits  blau_gender  Gini_gh_ten  user_id  Core1  commits  \
0          22688        0.000     0.422259  1614129      1    22138   
1          22688        0.000     0.422259  1614101      0      124   
2          22688        0.000     0.422259    94412      0       53   
3          22688        0.000     0.422259    88459      0       49   
4          22688        0.000     0.422259   738787      0       45   
..           ...          ...          ...      ...    ...      ...   
468         1655        0.153     0.292578   938914      0        1   
469         1224        0.000     0.229926    11648      0        1   
470         1224        0.000     0.229926   179354      0        1   
471         1224        0.000     0.229926   438642      0        1   
472         1224        0.000     0.229926     5268      0        1   

     propCommits gender  github_tenure  leavesNextQ  
0       0.975758   male            426            0  
1       0.005465   male             46            0  
2       0.002336   male            941            1  
3       0.002160   male           1870            0  
4       0.001983   male            781            0  
..           ...    ...            ...          ...  
468     0.000604   male            326            1  
469     0.000817   male            893            0  
470     0.000817   male           1238            0  
471     0.000817   male           1582            0  
472     0.000817   male           1903            0  

[473 rows x 16 columns]

# Rename Columns

In [4]:
# Rename columns as directed and save as new DataFrame to preserve history.
gh_tfp_renamed = (gh_turnover_four_projects
        .rename(columns={'domain':'project_domain', 
                         'language':'project_language',
                         'windows':'project_age',
                         'window_idx':'quarter',
                         'num_team':'team_size',
                         'num_commits':'project_commits',
                         'blau_gender':'gender_blau',
                         'Gini_gh_ten':'gh_tenure_gini',
                         'Core1':'core_dev',
                         'commits':'user_commits',
                         'propCommits':'commit_proportion',
                         'github_tenure':'gh_tenure',
                         'leavesNextQ':'leaves_next_q'})
)

# Quick inspection:
gh_tfp_renamed

project_id project_domain project_language  project_age  quarter  \
0          1101          OTHER           Python           23       22   
1          1101          OTHER           Python           23       22   
2          1101          OTHER           Python           23       22   
3          1101          OTHER           Python           23       22   
4          1101          OTHER           Python           23       22   
..          ...            ...              ...          ...      ...   
468       24293      DEV_FRAME           Python           24       23   
469       24293      DEV_FRAME           Python           24       24   
470       24293      DEV_FRAME           Python           24       24   
471       24293      DEV_FRAME           Python           24       24   
472       24293      DEV_FRAME           Python           24       24   

     team_size  project_commits  gender_blau  gh_tenure_gini  user_id  \
0           26            22688        0.000        0.422259  1614129   
1           26            22688        0.000        0.422259  1614101   
2           26            22688        0.000        0.422259    94412   
3           26            22688        0.000        0.422259    88459   
4           26            22688        0.000        0.422259   738787   
..         ...              ...          ...             ...      ...   
468         12             1655        0.153        0.292578   938914   
469         12             1224        0.000        0.229926    11648   
470         12             1224        0.000        0.229926   179354   
471         12             1224        0.000        0.229926   438642   
472         12             1224        0.000        0.229926     5268   

     core_dev  user_commits  commit_proportion gender  gh_tenure  \
0           1         22138           0.975758   male        426   
1           0           124           0.005465   male         46   
2           0            53           0.002336   male        941   
3           0            49           0.002160   male       1870   
4           0            45           0.001983   male        781   
..        ...           ...                ...    ...        ...   
468         0             1           0.000604   male        326   
469         0             1           0.000817   male        893   
470         0             1           0.000817   male       1238   
471         0             1           0.000817   male       1582   
472         0             1           0.000817   male       1903   

     leaves_next_q  
0                0  
1                0  
2                1  
3                0  
4                0  
..             ...  
468              1  
469              0  
470              0  
471              0  
472              0  

[473 rows x 16 columns]

# Data Wrangling

## Calculate the mean of `user_commits`. 

In [5]:
# Calculate by dividing the sum of 'user_commits' by the number of rows in the DataFrame.
sum(gh_tfp_renamed.user_commits) / len(gh_tfp_renamed)


339.36575052854124

In [6]:
# Or! Use the NumPy mean() function. This is much easier. See reference. 
np.mean(gh_tfp_renamed.user_commits)

339.36575052854124

## Calculate the median of `user_commits`.

In [7]:
# Calculate the median using the NumPy median() function. Neat!
np.median(gh_tfp_renamed.user_commits)

18.0

## Select all columns that start with a *p* (i.e., `(^p)`) **or** contain an *g* (i.e., `(g)`). Save it as a new DataFrame named `turnover_new`. Output the columns in your notebook.

In [8]:
# Select columns as instructed using .filter() and save as new DataFrame titled turnover_new.
turnover_new = (gh_tfp_renamed
 .filter(regex = '(^p)|(g)')
)

# Output the columns.
turnover_new.columns

Index(['project_id', 'project_domain', 'project_language', 'project_age',
       'project_commits', 'gender_blau', 'gh_tenure_gini', 'gender',
       'gh_tenure'],
      dtype='object')

## Using your newly created DataFrame `turnover_new`, select rows in which `project_commits` is greater than 9,000.


In [17]:
# Use the .query() function to get project_commits greater than 9000.
(turnover_new
 .query("project_commits > 9000")
)

project_id project_domain project_language  project_age  project_commits  \
0          1101          OTHER           Python           23            22688   
1          1101          OTHER           Python           23            22688   
2          1101          OTHER           Python           23            22688   
3          1101          OTHER           Python           23            22688   
4          1101          OTHER           Python           23            22688   
..          ...            ...              ...          ...              ...   
116        1101          OTHER           Python           23            14210   
117        1101          OTHER           Python           23            14210   
118        1101          OTHER           Python           23            14210   
119        1101          OTHER           Python           23            14210   
120        1101          OTHER           Python           23            14210   

     gender_blau  gh_tenure_gini gender  gh_tenure  
0            0.0        0.422259   male        426  
1            0.0        0.422259   male         46  
2            0.0        0.422259   male        941  
3            0.0        0.422259   male       1870  
4            0.0        0.422259   male        781  
..           ...             ...    ...        ...  
116          0.0        0.411372   male        495  
117          0.0        0.411372   male        487  
118          0.0        0.411372   male        317  
119          0.0        0.411372   male        418  
120          0.0        0.411372   male        179  

[114 rows x 9 columns]

## Using pandas piping notation, perform the previous two operations together and save it as a new DataFrame `turnover_newer`. This means you should select columns that start with a *p* or contain an *g* and select rows in which `project_commits` is greater than 9,000.

In [10]:
# Combine the previous two operations and save as turnover_newer.
turnover_newer = (gh_tfp_renamed
 .filter(regex = '(^p)|(g)')
 .query("project_commits > 9000")
)

# Check our work. They seem to be the same. 
turnover_newer

project_id project_domain project_language  project_age  project_commits  \
0          1101          OTHER           Python           23            22688   
1          1101          OTHER           Python           23            22688   
2          1101          OTHER           Python           23            22688   
3          1101          OTHER           Python           23            22688   
4          1101          OTHER           Python           23            22688   
..          ...            ...              ...          ...              ...   
116        1101          OTHER           Python           23            14210   
117        1101          OTHER           Python           23            14210   
118        1101          OTHER           Python           23            14210   
119        1101          OTHER           Python           23            14210   
120        1101          OTHER           Python           23            14210   

     gender_blau  gh_tenure_gini gender  gh_tenure  
0            0.0        0.422259   male        426  
1            0.0        0.422259   male         46  
2            0.0        0.422259   male        941  
3            0.0        0.422259   male       1870  
4            0.0        0.422259   male        781  
..           ...             ...    ...        ...  
116          0.0        0.411372   male        495  
117          0.0        0.411372   male        487  
118          0.0        0.411372   male        317  
119          0.0        0.411372   male        418  
120          0.0        0.411372   male        179  

[114 rows x 9 columns]

For the following tasks, *do not use* `turnover_new` or `turnover_newer`. Instead, use the original DataFrame you initially created from the provided file.

## Using pandas piping notation, select all columns that end with the letter *e* (i.e., `e$`) or contain the letter *s* (i.e., `s`). Additionally, select rows in which `gh_tenure` is greater than or equal to 1,991. 

In [11]:
# Select columns using .filter() and rows using .query().
(gh_tfp_renamed
 .filter(regex = '(e$)|(s)')
 .query("gh_tenure >= 1991")
)

project_language  project_age  team_size  project_commits  user_id  \
339             Java           16          2                7    33272   
457           Python           23         20             3070    88459   

     user_commits  gh_tenure  leaves_next_q  
339             6       2006              0  
457             1       2054              0

## Use the query you just performed. You will calculate the mean and median of `user_commits`. How does it compare to the answers above in which you calculated the mean and median? Please provide your answer in a Markdown cell within your notebook.

In [12]:
# Save query just performed as new DataFrame.
gh_tfp_renamed_2 = (gh_tfp_renamed
 .filter(regex = '(e$)|(s)')
 .query("gh_tenure >= 1991")
)

In [13]:
# Calulate mean.
np.mean(gh_tfp_renamed_2.user_commits)

3.5

In [18]:
# Calculate median. 
np.median(gh_tfp_renamed_2.user_commits)

3.5

In the previous calculation the mean was 339.366 and the median was 18. Those calculations took into account 473 projects and included an outlier with 22138 commits, the next closest being 9382 (seen in the sorted DataFrame below). The calculation immediatly above includes only two projects with much, much lower numbers of commits. With only two values, the mean and median are the same. 

In [15]:
(gh_tfp_renamed
.sort_values(by = ["user_commits"], ascending = False)
.head()
)

project_id project_domain project_language  project_age  quarter  \
0         1101          OTHER           Python           23       22   
26        1101          OTHER           Python           23        1   
34        1101          OTHER           Python           23        2   
35        1101          OTHER           Python           23        2   
47        1101          OTHER           Python           23       10   

    team_size  project_commits  gender_blau  gh_tenure_gini  user_id  \
0          26            22688          0.0        0.422259  1614129   
26          8             9939          0.0        0.314958    88459   
34         13            19010          0.0        0.251245   117614   
35         13            19010          0.0        0.251245    88459   
47         10             9280          0.0        0.385701     7245   

    core_dev  user_commits  commit_proportion gender  gh_tenure  leaves_next_q  
0          1         22138           0.975758   male        426              0  
26         1          9382           0.943958   male         16              0  
34         0          8207           0.431720   male        635              0  
35         0          7894           0.415255   male         67              0  
47         1          7504           0.808621   male         92              1

# References

The following website was used to learn how to used the NumPy .mean() and .median() funtions:

https://numpy.org/doc/2.2/reference/routines.statistics.html